### PR TITLE
Fixed cursor blinking during selection.

### PR DESCRIPTION
**Pull request by yiiim**
Fixed the issue of cursor blinking during selection as shown below.

https://github.com/flutter/flutter/assets/12887926/6f2caa9b-1748-4e1d-bbf0-3643793ad17c

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 1]
```diff
@@ -3215,6 +3215,10 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
         _lastTextPosition = currentTextPosition;
         renderEditable.setFloatingCursor(point.state, _lastBoundedOffset!, _lastTextPosition!);
       case FloatingCursorDragState.Update:
+        // Stop cursor blinking and making it visible.
+        _stopCursorBlink(resetCharTicks: false);
+        _cursorBlinkOpacityController.value = 1.0;
```
Comment (Renzo-Olivares) [inline packages/flutter/lib/src/widgets/editable_text.dart:3220]: It looks like we already call this on `FloatingCursorDragState.Start`, is something else resetting the cursor blink?

Comment (yiiim) [reply packages/flutter/lib/src/widgets/editable_text.dart:3220]: Yes, the cursor blinking is reset [here](https://github.com/flutter/flutter/blob/5887d6c74248fc6542a49e4a999b09b2aa58ef9c/packages/flutter/lib/src/widgets/editable_text.dart#L4041C39-L4041C39) when the selection changes. I should note this in the code.

Comment (Renzo-Olivares) [reply packages/flutter/lib/src/widgets/editable_text.dart:3221]: Thanks for the explanation. I think given these details, we should probably try prevent the blinking cursor from turning itself back on while the selection is changing due to floating cursor. Maybe by updating the logic in `_startOrStopCursorTimerIfNeeded`.

Comment (yiiim) [reply packages/flutter/lib/src/widgets/editable_text.dart]: I have made changes to the implementation, could you please take another look?

File: packages/flutter/lib/src/rendering/editable.dart [hunk 2]
```diff
@@ -1183,6 +1183,8 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin,
   /// Defaults to a padding with left, top and right set to 4, bottom to 5.
   EdgeInsets floatingCursorAddedMargin;
 
+  /// Returns true if the floating cursor is visible, false otherwise.
+  bool get floatingCursorOn => _floatingCursorOn;
```
Comment (Renzo-Olivares) [inline packages/flutter/lib/src/rendering/editable.dart:1187]: nit: I think it's fine to keep this private unless there is a use-case that requires this to be public.

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 3]
```diff
@@ -4008,7 +4008,7 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
   }
 
   void _stopCursorBlink({ bool resetCharTicks = true }) {
-    _cursorBlinkOpacityController.value = 0.0;
+    _cursorBlinkOpacityController.value = renderEditable.floatingCursorOn ? 1.0 : 0.0;
```
Comment (Renzo-Olivares) [inline packages/flutter/lib/src/widgets/editable_text.dart:4011]: Why is this needed? Shouldn't `_stopCursorBlink` stop the cursor blink as expected without changes?

Comment (yiiim) [reply packages/flutter/lib/src/widgets/editable_text.dart:4011]: Even when the floating cursor is displayed, we cannot ignore the call to `_stopCursorBlink` in `_didChangeTextEditingValue`, which leads to some test failures. I haven't investigated the specific reasons yet, but I think it's feasible to always keep the cursor visible in `_stopCursorBlink` when the floating cursor appears. Should I investigate the reasons for the test failures or maintain the status quo?

Comment (yiiim) [reply packages/flutter/lib/src/widgets/editable_text.dart:4011]: I think I've found some clues. We need to set the value of `_cursorBlinkOpacityController` to stop the animation. This is to avoid re-executing `_onCursorTick` when the animation ends at this location: https://github.com/flutter/flutter/blob/5887d6c74248fc6542a49e4a999b09b2aa58ef9c/packages/flutter/lib/src/widgets/editable_text.dart#L4004.

Comment (Renzo-Olivares) [reply packages/flutter/lib/src/widgets/editable_text.dart:4011]: Sorry for the delay on the review, just ran this locally and I understand what you mean. I'm okay with the code as-is then, but could you add a comment explaining why we take the floating cursor into account.

Comment (yiiim) [reply packages/flutter/lib/src/widgets/editable_text.dart:4019]: Done.

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 4]
```diff
@@ -4014,7 +4014,9 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
   }
 
   void _stopCursorBlink({ bool resetCharTicks = true }) {
-    _cursorBlinkOpacityController.value = 0.0;
+    // If the cursor is animating, stop the animation, and we always
+    // want the cursor to be visible when the floating cursor is enabled
```
Comment (Renzo-Olivares) [inline packages/flutter/lib/src/widgets/editable_text.dart:4018]: nit: add period at the end of this comment.

**Review by Renzo-Olivares — APPROVED** [review]
LGTM with small nit. Thank you for the contribution!

**Review by justinmc — APPROVED** [review]
LGTM 👍 

Good catch, thanks for fixing this! Is there an issue anywhere about this that you're aware of?

**Comment by yiiim** [comment]
> LGTM 👍
> 
> Good catch, thanks for fixing this! Is there an issue anywhere about this that you're aware of?

There is no related issue, this is a problem found after my other PR#138479.
